core/tests/test_utils.py

```diff
@@ -797,353 +355,784 @@ def log_line(self, line):
+                defaultTestResult() method) and used instead.
+        """
+        memory_cache_services_stub = MemoryCacheServicesStub()
+        memory_cache_services_stub.flush_cache()
+
+        with contextlib2.ExitStack() as stack:
+            stack.enter_context(self.swap(
+                platform_taskqueue_services, 'create_http_task',
+                self._taskqueue_services_stub.create_http_task))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'flush_cache',
+                memory_cache_services_stub.flush_cache))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'get_multi',
+                memory_cache_services_stub.get_multi))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'set_multi',
+                memory_cache_services_stub.set_multi))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'get_memory_cache_stats',
+                memory_cache_services_stub.get_memory_cache_stats))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'delete_multi',
+                memory_cache_services_stub.delete_multi))
+
+            super(AppEngineTestBase, self).run(result=result)
 
-        return self._send_post_request(
-            app, incoming_email_url, data,
-            expect_errors, headers=headers,
-            expected_status_int=expected_status_int)
+    def setUp(self):
+        self.testbed = testbed.Testbed()
+        self.testbed.activate()
+        self.testbed.setup_env(overwrite=True)
 
-    def post_task(
-            self, url, payload, headers, csrf_token=None, expect_errors=False,
-            expected_status_int=200):
-        """Posts an object to the server by JSON with the specific headers
-        specified; return the received object.
+        # Declare any relevant App Engine service stubs here.
+        self.testbed.init_user_stub()
+        self.testbed.init_app_identity_stub()
+        self.testbed.init_memcache_stub()
+        self.testbed.init_blobstore_stub()
+        self.testbed.init_urlfetch_stub()
+        self.testbed.init_files_stub()
+        self.testbed.init_search_stub()
+
+        policy = (
+            datastore_services.make_instantaneous_global_consistency_policy())
+        self.testbed.init_datastore_v3_stub(consistency_policy=policy)
+
+        # The root path tells the testbed where to find the queue.yaml file.
+        self.testbed.init_taskqueue_stub(root_path=os.getcwd())
+        self._taskqueue_stub = (
+            self.testbed.get_stub(testbed.TASKQUEUE_SERVICE_NAME))
+
+        # Set up the app to be tested.
+        self.testapp = webtest.TestApp(main.app)
+        self.taskqueue_testapp = webtest.TestApp(main_taskqueue.app)
+        self.mail_testapp = webtest.TestApp(main_mail.app)
+
+        self.signup_superadmin_user()
+
+    def tearDown(self):
+        datastore_services.delete_multi(
```

Comment:
`logout` has been removed from here because `testbed.deactivate` already does the job of restoring the os.environ.

Management of `os.environ` is an implementation detail of testbed to begin with, so we should not be managing it ourselves.